packages/cojson/src/coValues/coPlainText.ts

```diff
@@ -88,33 +83,22 @@ export class RawCoPlainText<
     text: string,
     privacy: "private" | "trusting" = "private",
   ) {
-    const ops: InsertionOpPayload<string>[] = [];
-    let prevOpId: OpID | "start" | undefined = this.mapping.opIDbeforeIdx[idx];
-    if (!prevOpId) {
-      if (idx === 0) {
-        prevOpId = "start";
-      } else {
-        throw new Error("Invalid idx");
+    const graphemes = [...this._segmenter.segment(text)].map((g) => g.segment);
+
+    if (idx === 0) {
+      // For insertions at start, just prepend each character, in reverse
+      for (const grapheme of graphemes.reverse()) {
+        this.prepend(grapheme, 0, privacy);
+      }
+    } else {
+      // For other insertions, use append after the specified index
+      // We append in forward order to maintain the text order
+      let after = idx - 1;
+      for (const grapheme of graphemes) {
+        this.append(grapheme, after, privacy);
+        after++; // Move the insertion point forward for each grapheme
       }
     }
```

Comment:
We can try to optimize this using `appendItems` which creates a single transaction.

It should also handle the "reverse" part, so it might help on simplifying the code here

Reply:
Yes, that makes sense, I'll give that a go

Reply:
Actually it doesn't port over, but I think that's a defect in `appendItems` I'll see if I can fix it